tests/swarm-deploy/test_swarm_runs.py

```diff
@@ -118,37 +116,73 @@ def test_core_service_running(
     # ID                  NAME                     IMAGE                     NODE                DESIRED STATE       CURRENT STATE            ERROR                       PORTS
     # puiaevvmtbs1        simcore_storage.1       simcore_storage:latest   crespo-wkstn        Running             Running 18 minutes ago
     # j5xtlrnn684y         \_ simcore_storage.1   simcore_storage:latest   crespo-wkstn        Shutdown            Failed 18 minutes ago    "task: non-zero exit (1)"
-    tasks = service.tasks()
-    num_tasks = get_replicas(service_config)
-
-    assert len(tasks) == num_tasks, (
-        f"Expected a {num_tasks} task(s) for '{service_name}', got instead"
-        f"\n{ get_tasks_summary(tasks) }"
-        f"\n{ get_failed_tasks_logs(service, docker_client) }"
-    )
 
-    for i in range(num_tasks):
-        task: Dict[str, Any] = service.tasks()[i]
+    expected_replicas = get_replicas(service_config)
 
-        for n in range(RETRY_COUNT):
-            task = service.tasks()[i]
+    # assert number of slots equals number of expected services
+    tasks = service.tasks()
+    slots: Set[int] = {t["Slot"] for t in tasks}

```

Comment:
basically if we want to have 8 sidecars, there should be 8 tasks, not anything else. If that is the case then we have something broken in the sidecar that needs fixing.

Reply:
### First of all, let me address the tasks and slots difference.

Each time a new service is started a new task for it is created for each replica. Each task has a unique slot assigned to it. This slot is an integer (usually you see it in the task name).
What they never say in the documentation (and is not obvious at all), if a task fails (regardless the of the reason), there will be multiple entries for the tasks with the same slot id.

To better explain see the below example. Start by defining a service with an image which dose not exist.

```bash
docker service create --replicas 1 --name aservice missingimage:latest2
```

To understand what is happening you actually need to use the `--no-trunc` option like below. Wait a few seconds before running the below command:
```bash
docker service ps --no-trunc aservice
```
The output will be similar to the following:
```
ID                          NAME             IMAGE                  NODE         DESIRED STATE   CURRENT STATE             ERROR                                   PORTS
44oiom4ba6rceyydxclrvaf9h   aservice.1       missingimage:latest2   neagu-wkst   Ready           Rejected 2 seconds ago    "No such image: missingimage:latest2"
lh15jmwbgmdj1k9xs2zugc8mj    \_ aservice.1   missingimage:latest2   neagu-wkst   Shutdown        Rejected 7 seconds ago    "No such image: missingimage:latest2"
lo0e312k39se79ljdi8d9m1v5    \_ aservice.1   missingimage:latest2   neagu-wkst   Shutdown        Rejected 12 seconds ago   "No such image: missingimage:latest2"
qzfnxfskeqq0swqvrte57ikxm    \_ aservice.1   missingimage:latest2   neagu-wkst   Shutdown        Rejected 17 seconds ago   "No such image: missingimage:latest2"
bt7z3ykrn3r9y08t37qrwgtwi    \_ aservice.1   missingimage:latest2   neagu-wkst   Shutdown        Rejected 22 seconds ago   "No such image: missingimage:latest2"
```

The information is not all that clear, but it goes as follows:
- the second line actually refers to the service itself and it states in what state the service is desired to be 
- lines 3-6 all contain a Task named `\_ aservice.1`, but all have different IDs, each of them complains that the image is missing (which is to be expected)
- so lines 3-6 actually refer to the same Task and have the same slot number but are in different states.
- when the tasks are recovered `tasks = service.tasks()` all the above will be returned, in this case, 4 tasks will be returned
- Currently, in the CI, when Posgress is unreachable there are 2 (or more) entries for the same task and slot id. One or more are in "Failed" and one is in "Running". Because the order of in which these tasks are listed is not guaranteed, you get random flaky CI tests and get to restart the CI countless times.

---

In conclusion:
- you can have more than one task listed `tasks = service.tasks()` for the same task/slot
- if there are 8 replicas, there might be more than 8 tasks listed

Reply:
I need to rename some of the variables inside the function for clarity, some of them are not very well named

Reply:
ok, so I think there is a small confusion there.
I did exactly as what you said but replaced replicas=1 by replicas=4:

and I obtained this (by the way, the no-trunc param only changes the ID from short to long so it makes no difference):
```bash
docker service ps aservice
ID             NAME             IMAGE                 NODE                   DESIRED STATE   CURRENT STATE             ERROR                              PORTS
y3zjnendm6ip   aservice.1       missingimage:latest   sanderegg-All-Series   Ready           Rejected 2 seconds ago    "No such image: missingimage:l…"   
tm4enmgl81za    \_ aservice.1   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 6 seconds ago    "No such image: missingimage:l…"   
ie904nrd8gll    \_ aservice.1   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 12 seconds ago   "No such image: missingimage:l…"   
gmfelojsodpq    \_ aservice.1   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 14 seconds ago   "No such image: missingimage:l…"   
w6ngk0fjtclo   aservice.2       missingimage:latest   sanderegg-All-Series   Ready           Rejected 2 seconds ago    "No such image: missingimage:l…"   
u5vxw9lnkldv    \_ aservice.2   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 6 seconds ago    "No such image: missingimage:l…"   
pilyh86flkgj    \_ aservice.2   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 12 seconds ago   "No such image: missingimage:l…"   
fg8kfoekjkwy    \_ aservice.2   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 14 seconds ago   "No such image: missingimage:l…"   
z68h4c0d4e3p   aservice.3       missingimage:latest   sanderegg-All-Series   Ready           Rejected 2 seconds ago    "No such image: missingimage:l…"   
qv8ucjr9edji    \_ aservice.3   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 7 seconds ago    "No such image: missingimage:l…"   
yvl4u4y8z41h    \_ aservice.3   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 12 seconds ago   "No such image: missingimage:l…"   
yto0eghef6un    \_ aservice.3   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 14 seconds ago   "No such image: missingimage:l…"   
4m8obbbjbob3   aservice.4       missingimage:latest   sanderegg-All-Series   Ready           Rejected 2 seconds ago    "No such image: missingimage:l…"   
cjamc81eot17    \_ aservice.4   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 6 seconds ago    "No such image: missingimage:l…"   
h6xeuyaj13ww    \_ aservice.4   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 12 seconds ago   "No such image: missingimage:l…"   
dn45522r2ysx    \_ aservice.4   missingimage:latest   sanderegg-All-Series   Shutdown        Rejected 14 seconds ago   "No such image: missingimage:l…"
```

Doing so effectively creates a service with 4 tasks. that is what we can see as aservice.1, aservice.2, aservice.3 and aservice.4. Respectively on lines where there is no ```\_``` sign.
The lines with the ```\_``` signs are the history and are shutdown. Always. This is the same that you see in the Portainer actually.

So now to come back to that test. we normally have 8 replicas of the sidecar. so there should be **exactly** 8 tasks not more not less. if there more, this means the first ones failed to start and this is something we would like to avoid since we want to have a controlled startup.